src/responder/kcm/kcmsrv_op_queue.c

```diff
@@ -75,6 +77,8 @@ struct kcm_ops_queue_ctx *kcm_ops_queue_create(TALLOC_CTX *mem_ctx)
         return NULL;
     }
 
+    queue_ctx->kctx = mem_ctx;
+
```

Comment:
You need to pass kctx directly to the function instead of abusing mem_ctx as that may be really anything.

Reply:
Kctx is the only argument passed to this function, does it mean the caller should be something like this?
~~~
kctx->qctx = kcm_ops_queue_create(kctx, kctx);
~~~
If yes, then how do I retrieve kctx in test_kcm_queue.c to pass it as an argument here? https://github.com/SSSD/sssd/blob/edec0ee3100ecb22958572d073ec635a9135037a/src/tests/cmocka/test_kcm_queue.c#L155

Reply:
The definition should be either `TALLOC_CTX *mem_ctx, struct kcm_ctx *kctx` or just `struct kcm_ctx *kctx`. Both are fine. But I'm afraid you will have to mock the structure in tests. See [mock_nctx] for example.

[mock_nctx]: https://github.com/SSSD/sssd/blob/master/src/tests/cmocka/test_nss_srv.c#L66

Reply:
I mocked the `kcm_ctx` structure in the tests(see updated code here in the PR) but the test crashes here. Do I need to mock the `rctx` as well to address this? It appears that is a bit more involved.

~~~
Program received signal SIGSEGV, Segmentation fault.
0x0000000000403074 in kcm_op_queue_entry_destructor (entry=0x40f2c0) at ../src/responder/kcm/kcmsrv_op_queue.c:132
132         } else if (terminating) {
(gdb) l
127         terminating = entry->queue->qctx->kctx->rctx->shutting_down;
128
129         if (entry == NULL) {
130             return 1;
131         /* Prevent use-after-free of req when shutting down with non-empty queue */
132         } else if (terminating) {
133             return 0;
134         }
135
136         /* Take the next entry from the queue */
(gdb) where
#0  0x0000000000403074 in kcm_op_queue_entry_destructor (entry=0x40f2c0) at ../src/responder/kcm/kcmsrv_op_queue.c:132
#1  0x00007ffff7903d26 in _tc_free_internal () from /lib64/libtalloc.so.2
#2  0x00007ffff79039d8 in _tc_free_internal () from /lib64/libtalloc.so.2
#3  0x00007ffff7916df5 in tevent_req_received () from /lib64/libtevent.so.0
#4  0x00007ffff7916e0d in tevent_req_destructor () from /lib64/libtevent.so.0
#5  0x00007ffff7903d26 in _tc_free_internal () from /lib64/libtalloc.so.2
#6  0x0000000000402dac in test_kcm_queue_done (req=0x40eb40) at ../src/tests/cmocka/test_kcm_queue.c:193
#7  0x00007ffff791af7d in tevent_common_invoke_timer_handler () from /lib64/libtevent.so.0
#8  0x00007ffff791b120 in tevent_common_loop_timer_delay () from /lib64/libtevent.so.0
#9  0x00007ffff791c4e9 in epoll_event_loop_once () from /lib64/libtevent.so.0
#10 0x00007ffff791a57b in std_event_loop_once () from /lib64/libtevent.so.0
#11 0x00007ffff7915598 in _tevent_loop_once () from /lib64/libtevent.so.0
#12 0x0000000000402bb6 in test_kcm_queue_single (state=<optimized out>) at ../src/tests/cmocka/test_kcm_queue.c:239
#13 0x00007ffff7fa3d31 in cmocka_run_one_test_or_fixture () from /lib64/libcmocka.so.0
#14 0x00007ffff7fa4465 in _cmocka_run_group_tests () from /lib64/libcmocka.so.0
#15 0x00000000004026ee in main (argc=1, argv=<optimized out>) at ../src/tests/cmocka/test_kcm_queue.c:379
(gdb) p entry->queue->qctx->kctx
$1 = (struct kcm_ctx *) 0x40e740
(gdb) p entry->queue->qctx->kctx->rctx
$2 = (struct resp_ctx *) 0x0
~~~

Reply:
You're relying on `entry->queue->qctx->kctx->rctx->shutting_down` so you also need to mock `rctx`. This is already available in `src/tests/cmocka/common_mock_resp.c:mock_rctx`.

Reply:
Do you know why I see the linking errors when adding `#include "tests/cmocka/common_mock_resp.h"` to test_kcm_queue.c and running `make check` ?
~~~
/usr/bin/ld: src/tests/cmocka/test_kcm_queue-test_kcm_queue.o: in function `setup_kcm_queue':                                                                                                                                                 
/home/justin/github/sssd/x86_64/../src/tests/cmocka/test_kcm_queue.c:171: undefined reference to `mock_rctx'                                                                                                                                  
collect2: error: ld returned 1 exit status     
make[3]: *** [Makefile:18593: test_kcm_queue] Error 1
make[3]: Leaving directory '/home/justin/github/sssd/x86_64'
make[2]: *** [Makefile:39905: check-am] Error 2
make[2]: Leaving directory '/home/justin/github/sssd/x86_64'
make[1]: *** [Makefile:38814: check-recursive] Error 1
make[1]: Leaving directory '/home/justin/github/sssd/x86_64'
make: *** [Makefile:39908: check] Error 2
~~~

Reply:
You need to add `src/tests/cmocka/common_mock_resp.c` source file to [Makefile.am](https://github.com/SSSD/sssd/blob/master/Makefile.am#L3950)

Reply:
`mock_rctx` is defined in `common_mock_resp.c`

See https://github.com/SSSD/sssd/blob/master/Makefile.am#L3949

Depending on what you eventually need to link with, you can either add `common_mock_resp.c` directly to _SOURCES list or use existing `TEST_MOCK_RESP_OBJ` (but this is large so most probably not needed here)